openwisp_users/multitenancy.py

```diff
@@ -8,7 +8,23 @@
 OrganizationUser = load_model('openwisp_users', 'OrganizationUser')
 
 
-class MultitenantAdminMixin(object):
+class SelectOrgMixin(object):
+    def select_organization(self, request, form):
+        '''
+        When the user has access to only one organization
+        automatically set that org as the default value
+        '''
+        if request.POST:
+            return
+        fields = form.base_fields
+        if 'organization' in fields:
+            org_field = fields['organization']
+            org_queryset = org_field._queryset
+            if org_queryset.count() == 1:
+                org_field.initial = org_queryset[0]
```

Comment:
There's two problems here.

1. The system may have multiple organizations but the user has only access to one.
2. The query will be repeated in every page, but will yield always the same result for a given user

For 1, I'm not sure this code addresses this. Does it?
We surely need a test for this.

Alternatively to check that a nonsuper user has access to only 1 org we could use:
```python
if not request.user.is_superuser and len(request.user.organization_dict.keys()) == 1:
    # ...
```

But only I'd use it only if this code does not already solve it, it may be possible that the queryset is already preprocessed by the multitenancy admin mixin, would be good if you confirm this.

For problem 2., I think we should cache the query for each user and add caching invalidation, like we already do for organizations_dict:

- caching: https://github.com/openwisp/openwisp-users/blob/3e28715c94617faeab5d23514dfa3ccbbf217975/openwisp_users/base/models.py#L94-L120
- invalidation: https://github.com/openwisp/openwisp-users/blob/master/openwisp_users/apps.py#L86-L102

Reply:
In the invalidation when `update_organizations_dict` function is called?

Reply:
I have confirmed the queryset is already preprocessed by the multitenancy admin mixin and the code is solving 1st problem.

Reply:
@nemesisdesign  I have tried to cached the query using following code.

```
def select_organization(self, request, form):
        '''
        When the user has access to only one organization
        automatically set that org as the default value
        '''
        if request.POST:
            return
        fields = form.base_fields
        if 'organization' in fields:
            org_field = fields['organization']
            org_field.initial = request.user.selected_org(org_field)

```
and 
```
def selected_org(self,org_field):
        # print('here')
        # print(self.username)
        cache_key = 'selected_{}_org'.format(self.pk)
        selected_org = cache.get(cache_key)
        # print(selected_org)
        if selected_org is not None: # If cache is not deleted
            # print('cached')
            if selected_org == False:
                return None 
            return selected_org
        # print('not cached')
        org_queryset = org_field._queryset
        if org_queryset.count() == 1:
                selected_org = org_queryset[0]
        else:
            selected_org = False  # Tells selected org is None
        cache.set(cache_key, selected_org, 86400 * 2)  # Cache for two days
``` 
This in models.py of base.

But I am able to delete the cache with the help of invalidation as you have suggested because I can't get request.user in signal.

Reply:
I have added caching. Please review it once so that I can work on testing.